### PR TITLE
GA

### PR DESCRIPTION
**Pull request by nathanajah**

### CODE REVIEW

File: GA.java [hunk 1]
```diff
@@ -1,26 +1,91 @@
...
-	 * 
-	 * @param seed
+	 * Run some iterations of the genetic algorithm to refine the population.
+	 * @param iterations The number of iterations.
 	 */
-	public void run(int seed) {
-		
+	public void run(int iterations) {
+		Random rnd = new Random();
+    for (int i = 0; i < iterations; i++) {
```
Comment (digawp) [inline GA.java]: Indentation! Use IDE pls lol

File: GA.java [hunk 2]
```diff
@@ -1,26 +1,91 @@
-
+import java.util.Random;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
 public class GA extends Learning {
```
Comment (digawp) [inline GA.java:6]: Shouldn't we have a blank line after imports? (Sorry nitpicking lol)

File: GA.java [hunk 3]
```diff
@@ -1,26 +1,91 @@
...
   Strategy[] population;
```
Comment (digawp) [inline GA.java]: Indentation

File: GA.java [hunk 4]
```diff
@@ -1,26 +1,91 @@
...
-	
 	/**
-	 * 
-	 * @param seed
+	 * Run some iterations of the genetic algorithm to refine the population.
+	 * @param iterations The number of iterations.
 	 */
-	public void run(int seed) {
-		
+	public void run(int iterations) {
```
Comment (digawp) [inline GA.java:53]: yeah I guess this is a better name for the argument. Will change the abstract method signature on my branch

File: GA.java [hunk 5]
```diff
@@ -1,26 +1,91 @@
...
+			List<Strategy> newPopulation = new ArrayList<Strategy>();
+			for (int j = 0; j < NUM_NEW_POPULATION/2; j++) {
+				int idx1 = rnd.nextInt(NUM_POPULATION);
+				int idx2 = rnd.nextInt(NUM_POPULATION);
+				Strategy[] newStrats = population[idx1].crossover(population[idx2]);
+				for (int k = 0; k < 2; k++) {
+					newStrats[k].mutate();
+					newStrats[k].normalize();
+					newPopulation.add(newStrats[k]);
+					newPopulation.add(newStrats[k]);
```
Comment (digawp) [inline GA.java]: Add the new strategy 2 times? Or typo?

Comment (nathanajah) [inline GA.java]: Two strategies can be retrieved from one crossover (first part of A + second part of B, first part of B + second part of A)

Comment (digawp) [inline GA.java]: Yes but it is being handled by the for loop right? You are adding the same strategy twice here. (In total adding 4 strategies in this 4 loops, in which two of them are the same)

Comment (digawp) [inline GA.java]: Btw the indentation of this whole for loop block is off also....

File: GA.java (hunk 4 above)
Comment (digawp) [inline GA.java:53]: oh nvm I did not declare abstract method `run` lol

File: GA.java (hunk 5 above)
Comment (nathanajah) [inline GA.java]: I thought I fixed it already :o
Will fix.

**Pull request merged**